cmd/proxy/actions/filter.go

```diff
@@ -0,0 +1,60 @@
+package actions
+
+import (
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/bketelsen/buffet"
+	"github.com/gobuffalo/buffalo"
+	"github.com/gomods/athens/pkg/errors"
+	"github.com/gomods/athens/pkg/module"
+	"github.com/gomods/athens/pkg/paths"
+)
+
+func newFilterMiddleware(mf *module.Filter) buffalo.MiddlewareFunc {
+	const op errors.Op = "actions.FilterMiddleware"
+
+	return func(next buffalo.Handler) buffalo.Handler {
+		return func(c buffalo.Context) error {
+			sp := buffet.SpanFromContext(c).SetOperationName("filterMiddleware")
+			defer sp.Finish()
+			mod, err := paths.GetModule(c)
+
+			if err != nil {
+				// if there is no module the path we are hitting is not one related to modules, like /
+				return next(c)
+			}
+
+			// not checking the error. Not all requests include a version
+			// i.e. list requests path is like /{module:.+}/@v/list with no version parameter
+			version, _ := paths.GetVersion(c)
+
+			if isPseudoVersion(version) {
+				return next(c)
+			}
+
+			rule := mf.Rule(mod)
+			switch rule {
+			case module.Exclude:
+				return c.Render(http.StatusForbidden, nil)
+			case module.Private:
+				return next(c)
+			case module.Include:
+				// TODO : spin up cache filling worker and serve the request using the cache
```

Comment:
@arschles why do we need a worker instead of just calling `next(c)`?

Reply:
I'm not sure if we decided for sure, but iirc we were talking about redirecting to olympus and downloading into the local storage in the background? If I'm remembering right, that would be what the worker is for

Reply:
opened https://github.com/gomods/athens/issues/471 , if we want to next() we'll need to change the download handlers to hit olympus instead of the vcs

Reply:
I created https://github.com/gomods/athens/issues/474 to continue discussion